APITests/utils.py

```diff
@@ -61,250 +62,252 @@ def format(self, record):
-        ]
-        all_status_code = {"200": 0, "500": 0, "503": 0, "504": 0}
-        for f in concurrent.futures.as_completed(futures):
-            try:
-                status_code = f.result().status_code
-                status_code_str = str(status_code)
-                if status_code_str != "200":
-                    logger.error(
-                        f"{status_code} error running: {url} with using params {params}"
-                    )
-                all_status_code[status_code_str] = all_status_code[status_code_str] + 1
-            except InvalidSchema:
-                raise InvalidSchema(
-                    f"No connection adapters were found for {url}. Please make sure that your URL is correct. "
+        if request_type == "get":
+            cal_run_time = CalculateRunTime(url, headers)
+            futures = [
+                executor.submit(cal_run_time.fetch, params)
+                for x in range(concurrent_threads)
+            ]
+        else:
+            # post request requires a send_manifest function and a valid manifest file path
+            if not manifest_to_send_func or not file_path_manifest:
+                raise ValueError(
+                    "Please provide a function to send manifest and a valid manifest file path"
                 )
+            else:
+                futures = [
+                    executor.submit(manifest_to_send_func, file_path_manifest, params)
+                    for x in range(concurrent_threads)
+                ]
+    http_responses = []
+    for f in concurrent.futures.as_completed(futures):
+        try:
+            response = f.result()
+            http_responses.append(response)
+            if response.status_code != 200:
+                logger.error(
+                    f"{response.status_code} error running: {url} with using params {params}"
+                )
+        except InvalidSchema:
+            raise InvalidSchema(
+                f"No connection adapters were found for {url}. Please make sure that your URL is correct. "
+            )
+    return http_responses
 
-    time_diff = round(time.time() - start_time, 2)
-    logger.info(f"duration time of running {url}: {time_diff}")
-    return dt_string, time_diff, all_status_code
 
+def parse_http_requests(http_responses: Response) -> dict[str, int]:
+    """parse http requests and return a dictionary that contains string and integers
 
-def cal_time_api_call_post_request(
+    Args:
+        http_responses (Responses): http responses
+
+    Returns:
+        dict: a status dictionary
+    """
+    all_status_code = {"200": 0, "500": 0, "503": 0, "504": 0}
+    for response in http_responses:
+        status_code = response.status_code
+        status_code_str = str(status_code)
+
+        if status_code_str in all_status_code:
+            all_status_code[status_code_str] += 1
+        else:
+            logger.error("an error occurred")
+    return all_status_code
+
+
+def cal_time_api_call(
```

Comment:
@BryanFauble how does the code look for you now? I separated the code into "execution" step and "parsing" step like you recommended. Thanks again for your help!

Reply:
@linglp Take a look at this commit here:
https://github.com/Sage-Bionetworks/schematic_profiler/commit/96f95f9f5a50577e984c11dca0564cf3dea24015

This is NOT TESTED CODE.

The code is closer to my thoughts on structuring the code. The code you wrote works, but I feel like it is still too deep of a call tree where it's adding more complexity than required. In your code the call chain is:
```
APITests/annotations_upload_submission_benchmark.py#execute_submission_comparison
->
CalculateRunTime constructor
->
send_request -> cal_time_api_call -> [execute_requests, parse_http_requests] -> [cal_run_time.fetch or submit manifest_to_send_func]

<- return
```

This is a deeply nested chain of calls:
```
send_request -> cal_time_api_call -> [execute_requests, parse_http_requests] -> [cal_run_time.fetch or submit manifest_to_send_func]
```
The problem here is that there is logic on what is actually being executed in 2 different spots. once in `execute_submission_comparison` where you're specifying you want to execute `send_manifest` and secondly inside of `execute_requests` where you might instead use `cal_run_time.fetch`.


Conversely in my suggestion, the call tree is flatter:

```
APITests/annotations_upload_submission_benchmark.py#execute_submission_comparison
->
perf_counter
->
create send_manifest_async tasks
->
execute_requests_async -> context manager/for loops with asyncio
->
parse_http_requests
->
find time diff

<- return
```

Reply:
Also:
```

async def execute_submission_comparison(
    dataset_id: str, asset_view_id: str, file_path_manifest: str, num_time: int
) -> None:
    """for this use case, only interested in if file_annotations_upload_lst is set to True or False

    Args:
        dataset_id (str): dataset id on synapse
        asset_view_id (str): asset view id on synapse
        file_path_manifests (Path): path of test manifest for submission
        num_time (int): number of times that submission function runs
    """
    srt = StoreRuntime()
    token = srt.get_access_token()
    headers = {"Authorization": f"Bearer {token}"}
    file_annotations_upload_lst = [True, False]
    base_url = "http://localhost:3001/v1/model/submit"
    params = {
        "manifest_record_type": "file_only",
        "schema_url": "https://raw.githubusercontent.com/Sage-Bionetworks/schematic/develop/tests/data/example.model.jsonld",
        "dataset_id": dataset_id,
        "asset_view": asset_view_id,
        "restrict_rules": False,
        "use_schema_label": "class_label",
        "data_model_labels": "class_label",
    }

    for i in file_annotations_upload_lst:
        submission_duration = []
        params["file_annotations_upload"] = i

        # try submitting x amount of time and calculate average
        for _ in range(num_time):
            start_time = perf_counter()
            requests_to_execute = []
            for i in range(1):
                requests_to_execute.append(
                    asyncio.create_task(
                        send_manifest_async(
                            url=base_url,
                            headers=headers,
                            params=params,
                            manifest_path=file_path_manifest,
                        )
                    )
                )
            http_responses = await execute_requests_async(requests_to_execute)
            all_status_code = parse_http_requests(http_responses)
            time_diff = round(perf_counter() - start_time, 2)
            if all_status_code["200"] != 1:
                print("encountered an error when submitting the manifest")
            else:
                submission_duration.append(time_diff)
        average_submission_time = sum(submission_duration) / len(submission_duration)
        logger.info(
            f"average time of submitting a manifest when upload_file_annotations set to {i} is: {average_submission_time}"
        )
```


Is just a thought. It's pulling out some of the ideas you had in the `cal_time_api_call` function. You could still have a `cal_time_api_call` type function out there. But instead it could just take in one or more futures to execute. Then you can write all:
```
            start_time = perf_counter()
            http_responses = await execute_requests_async(requests_to_execute)
            all_status_code = parse_http_requests(http_responses)
            time_diff = round(perf_counter() - start_time, 2)
```

Once and re-use it as the generic way to wrap the calls with something that'll capture the time diff and still return the 3 result tuples.

Reply:
Hi Bryan! Thank you so much for your feedbacks. Sorry it took me a while to think and understand the code. I made some changes to `cal_time_api_call`, and now the call tree looks like this: 
manifest_generate_benchmark.py(function `execute_manifest_generate_use_annotations_comparison`) -> cal_time_api_call -> execute_request

`execute_request` will now be responsible for executing requests in `await asyncio.gather`

Does that look closer to your idea? I tested and now the code is working. But I don't know how to incorporate threads with async calls or if that's still even necessary. The initial purpose of adding threading is to test if the application could handle concurrent requests. But I think that purpose could still be served with async calls?